### PR TITLE
Refactor OpenAI Models for parity with Transformers Continuation API

### PR DESCRIPTION
**Pull request by arunpatro**
I have started the refactor work based on our call. However I am kinda confused as how to proceed.

### CODE REVIEW

**Comment by rlouf** [comment]
I updated the PR. I don't think it will work as is, but it should give you an idea of what the design should look like.

**Comment by arunpatro** [comment]
This code works as intended. But pytest fails for `max_tokens` and `stop` becaues of the new api for continuation. However this code works:

```python
model = models.openai("gpt-4", temperature=0.7)
text.generate.continuation(model, max_tokens=100, stop=['tiger'])("This is a story about a tiger")
```

File: outlines/text/generate/continuation.py [hunk 1]
```diff
@@ -88,10 +91,16 @@ def postprocess_completions(self, completions: List[str]) -> List[str]:
         return completions_without_stop
 
 
-def continuation(
+@functools.singledispatch
+def continuation(model, max_tokens):
```
Comment (rlouf) [inline outlines/text/generate/continuation.py:95]: I believe `pytest` fails because this function needs to have the same signature as the other ones.

Comment (arunpatro) [reply outlines/text/generate/continuation.py:95]: I don't think so. Looks like the `Model()` class defined in the test doesn't match the types we have created in this dispatch. So it throws a `NotImplementedError``

**Pull request closed without merging**